### PR TITLE
Android: Fix playback getting stuck when refreshrate switch is enabled

### PR DESCRIPTION
**Pull request by a1rwulf**
## Description
When playing video and refreshrate switch is enabled, sometimes playback gets stuck a few seconds into playback.
The reason for this is that AudioTrack.write(...) never returns and thus our whole playback process stalls.
In order to not break AudioTrack, we need to make sure we suspend the AudioEngine early enough.

This change moves suspension of AE before we actually call the Android API to set the desired mode, instead of waiting that it already is done.

## Motivation and context
Fixes a nasty issue that sometimes happens and needs a reboot of Android to get fixed again.

## How has this been tested?
Tested with Nvidia Shield.

## What is the effect on users?
Working playback when refreshrate switch is on.

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by a1rwulf** [comment]
@fritsch that's the mentioned change.
2 aspects that I changed on purpose:
- suspend AE earlier, which seems to fix the issue I face a lot with certain HDMI equipment attached
- remove the isPlaying check, becasue it doesn't matter if we're playing, modechanges need AE suspended at all times

**Comment by thexai** [comment]
Just say that I have Shield Pro 2019 with switch refresh rate on and never I have experienced this.
I've been using it a lot lately with UHD Remuxes, HDR10 and Dolby Vision, audio passthrough, etc.

I'm not saying that the PR is wrong, just that it's not a general issue on Android or Shield.

**Comment by a1rwulf** [comment]
> Just say that I have Shield Pro 2019 with switch refresh rate on and never I have experienced this. I've been using it a lot lately with UHD Remuxes, HDR10 and Dolby Vision, audio passthrough, etc.
> 
> I'm not saying that the PR is wrong, just that it's not a general issue on Android or Shield.

True, I never faced this issue with my AVR either.
But I face it every 5-10th playback, when connected to HDMI Matrixes of different brands.
Seems to be timing sensitive issue, especially happening with non standard equipment (but might even happen with certain AVR/TV brands).

**Comment by Tourosi** [comment]
When I used the amlogic TV box, I fixed this issue in a similar way.
It didn't always happen in that box, and it didn't happen in Shield TV and FireTV.
I think your PR is very nice work, but I doubt this PR would be merged.

File: xbmc/platform/android/activity/XBMCApp.cpp [hunk 1]
```diff
@@ -722,18 +722,14 @@ void CXBMCApp::SetRefreshRate(float rate)
   }
 
   m_refreshRate = rate;
-
   m_displayChangeEvent.Reset();
+
+  if (m_hdmiSource)
+    dynamic_cast<CWinSystemAndroid*>(CServiceBroker::GetWinSystem())->InitiateModeChange();
```
Comment (lrusak) [inline xbmc/platform/android/activity/XBMCApp.cpp:728]: Can this be a static_cast?

**Review by fritsch — APPROVED** [review]
Thanks much. Let's try it that way.

Do you have any downside in mind, like devices where we don't get a hotplug?

**Comment by a1rwulf** [comment]
> Thanks much. Let's try it that way.
> 
> Do you have any downside in mind, like devices where we don't get a hotplug?

I don't think this implementation has a serious downside.
However it's a partial revert of: https://github.com/xbmc/xbmc/pull/15732 namely it reverts:
> 1.) Improve mode switch by not resetting Audio before android's new mode is triggered

I don't see how this even "improved" the modeswitch, it just increases the chance that you get into the mentioned fatal error situation.
As far as I understand the original intention was a micro optimization to reset audio a few ms later, but imho correct is to suspend AE immediately and not when Android has already triggered a modechange.

The only possible downside I can imagine is that on the mentioned AFTV 2018, when there is a modechange but no HDMI signal switch (I can't even tell how that's possible), you might have a few ms longer pause state due to the earlier AE suspend - I bet it's barely noticable, if at all.


I don't think this change has any impact regarding hotplug, the hotplug is related to when we reset display not when we lose display, this PR doesn't change the logic for OnResetDisplay - it just makes sure we have suspended AE before any mode change work is triggered.

**Pull request merged**